### PR TITLE
Adding is_palindrome and test_is_palindrome

### PR DESCRIPTION
**Pull request by Mykytako**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by suhrobmuboraksho — COMMENTED** [review]
Good job Mykyta. Did you try to test with no arguments, ie is_palindrome()

**Comment by Mykytako** [comment]
> Good job Mykyta. Did you try to test with no arguments, ie is_palindrome()

Thank you for your suggestion! Just made the changes

File: solutions/is_palindrome.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 2025-01-03
```
Comment (olumide-AI) [inline solutions/is_palindrome.py:4]: Your module docstring could be more informative Example, you could do something like this 
This module provides a function to check if a given string is a palindrome.

Module contents: that talks about usage and the function

File: solutions/tests/test_is_palindrome.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Unit tests for the is_palindrome function.
```
Comment (olumide-AI) [inline solutions/tests/test_is_palindrome.py:4]: Module docstring could be more descriptive, but good overall

File: solutions/tests/test_is_palindrome.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+Created on 2025-01-03
+@author: Your Name
+"""
+
+import unittest
+
+from solutions.is_palindrome import is_palindrome
+
+
+class TestPalindrome(unittest.TestCase):
```
Comment (olumide-AI) [inline solutions/tests/test_is_palindrome.py:14]: Your class should be TestIsPalindrome

File: solutions/tests/test_is_palindrome.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    def test_non_string_input(self):
+        """It should raise AssertionError when the input is not a string"""
+        with self.assertRaises(AssertionError) as context:
+            is_palindrome(12345)
+        self.assertEqual(str(context.exception), "Input should be a string")
+
+    def test_no_argument(self):
+        """It should return True when no argument is given (default empty string)"""
+        self.assertTrue(is_palindrome())
```
Comment (olumide-AI) [inline solutions/tests/test_is_palindrome.py:58]: you can add this to the end of the code, i don't really understand why exactly but evan used it in his code as well 
if __name__ == "__main__":
    unittest.main()

File: solutions/tests/test_is_palindrome.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+from solutions.is_palindrome import is_palindrome
+
+
+class TestPalindrome(unittest.TestCase):
+    """
+    This class contains unit tests for the is_palindrome function.
+    The function checks if a string is a palindrome.
+    """
+
+    def test_empty_string(self):
```
Comment (olumide-AI) [inline solutions/tests/test_is_palindrome.py:20]: For clarity you can separate your test and put comments in your code to show where each test falls into. Ex:
#Standard cases 
 test_simple, test_simple_non , test_case_insensitive
#defensive test
test_none_input
test_non_string imput 
#edge cases
- i think your others are edge cases

File: solutions/is_palindrome.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+"""
+
+
+def is_palindrome(s: str = "") -> bool:
+    """
+    Function to check if a string is a palindrome.
+    Input: s[str] (optional, default empty string)
+    Return: result[bool] True if the string is a palindrome, False otherwise
+
+    >>> is_palindrome('racecar')
```
Comment (olumide-AI) [inline solutions/is_palindrome.py:15]: You have too much docstring 3-4 is fine

File: solutions/is_palindrome.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 2025-01-03
+@author: Mykyta Kondratiev
+"""
+
+
+def is_palindrome(s: str = "") -> bool:
```
Comment (olumide-AI) [inline solutions/is_palindrome.py:9]: s is not a well descriptive name, maybe something like text, input_string could be better for readability

File: solutions/is_palindrome.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    True
+    >>> is_palindrome('hello')
+    False
+    >>> is_palindrome('')
+    True
+    >>> is_palindrome('A man a plan a canal Panama')
+    True
+    >>> is_palindrome('Was it a car or a cat I saw?')
+    True
+    >>> is_palindrome()
```
Comment (olumide-AI) [inline solutions/is_palindrome.py:25]: Line 25 fails when i try to run it, if could look into that. Not exactly sure why. Maybe it's expected. I you could pls explain why that will be nice to better understand the code. Thank you

Comment (olumide-AI) [reply solutions/is_palindrome.py:25]: Maybe do an assert raise or something like that?

**Review by olumide-AI — COMMENTED** [review]
Already great work, i highlighted a couple concerns on my single comment. I think it's ready for approval once you resolve our conversations. Thank you and i hope i helped. Overall this particular code was very difficult for me to understand for my level but i tried to give you as much constructive feedback as possible 



## Behavior

### Files

- [yes ] The file name describes the function's behavior
- [yes,could be better  There is a module docstring in the function file
- [yes] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [yes,could be better ] There is a module docstring in the tests file

### Unit Tests

- [yes ] The test class has a helpful name in PascalCase
- [yes ] The test class has a docstring
- Every unit test has
  - [yes ] A helpful name
  - [yes ] A clear docstring
  - [yes ] Only one assertion
  - [ yes] There is no logic in the unit test
- [yes] All tests pass
- [yes ] There are tests for defensive assertions
- [yes ] There are tests for boundary cases

### Function Docstring

- [yes ] The function's behavior is described
- The function's arguments are described:
  - [yes ] Type
  - [yes ] Purpose
  - [yes] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [yes] Type
  - [yes] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [yes ] Each assumption about an argument is checked with an assertion
  - [yes ] Each assertion checks for _only one_ assumption about the argument
- [yes] Include 3 or more (passing!) doctests

### The Function

- [yes ] The function's name describes it's behavior
- [yes ] The function's name matches the file name
- [yes ] The function has correct type annotations

## Strategy

### Do's

- [yes ] Variable names help to understand the strategy
- [yes ] Any comments are clear and describe the strategy
- [yes] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [pass ] The function's strategy _is not_ described in the documentation
- [ pass] Comments explain the _strategy_, **not** the _implementation_
- [ pass] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [yes ] The code passes the formatting checks
- [ yes] The code passes all Ruff linting checks
- [yes ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [yes ] Variables are named with snake_case
- [ yes] Variable names are clear and helpful
- [yes ] The code follows the strategy as simply as possible
- [yes ] The implementation is as simple as possible given the strategy
- [yes ] The code includes defensive assertions
- [ yes] Defensive assertions include as little logic as possible

**Review by suhrobmuboraksho — COMMENTED** [review]
Left some comments.

**Review by Mithchell509 — COMMENTED** [review]
Judicious, I'm inspired by it for my second solution

**Comment by Mykytako** [comment]
> Judicious, I'm inspired by it for my second solution

Thank you :) Could you approve the request please if everything is fine for you?

**Review by suhrobmuboraksho — APPROVED** [review]
Docstring is comprehensive.
No linting errors.
Comprehensive unittests.
Tests for edge cases.

**Review by yuri-spizhovyi-mit — APPROVED** [review]
Both the is_palindrome function and its test suite adhere to best practices and meet all criteria in the provided checklist. The solution is well-written, fully tested, and ready for production.

**Review by Mithchell509 — APPROVED** [review]
